ESP_ATMod/command.cpp

```diff
@@ -1742,105 +1746,68 @@ void cmd_AT_CIPSSLCERT()
 			Serial.printf_P(MSG_ERROR);
 		}
 	}
-	else if (inputBuffer[offset] == '?' && (inputBufferCnt >= 16 && inputBufferCnt <= 18))
+	// Print all certificates
+	else if (inputBuffer[offset] == '?' && inputBufferCnt == offset + 3)
 	{
 		if (CAcert.getCount() == 0)
 		{
-			Serial.println(F("+CIPSSLCERT:no cert"));
+			Serial.println(F("+CIPSSLCERT:no certs loaded"));
+			Serial.printf_P(MSG_ERROR);
 		}
 		else
 		{
-			uint32_t certNumber;
-
-			++offset;
-			if (!readNumber(inputBuffer, offset, certNumber))
-			{
-				certNumber = 1;
-			}
-
-			if (certNumber > CAcert.getCount())
+			for (size_t i = 0; i < CAcert.getCount(); i++)
 			{
-				Serial.println(F("+CIPSSLCERT:no cert"));
+				Serial.printf_P(PSTR("+CIPSSLCERT,%d:"), i + 1);
+				printCertificateName(i);
 			}
-			else
-			{
-				Serial.print(F("+CIPSSLCERT:"));
-
-				const br_x509_certificate *cert = &(CAcert.getX509Certs()[certNumber - 1]);
 
-				uint8_t *cnBytes = getCnFromDer(cert->data, cert->data_len);
-
-				if (cnBytes != nullptr)
-				{
-					char *cn = new char[cnBytes[0] + 1];
-
-					if (cn != nullptr)
-					{
-						memcpy(cn, &(cnBytes[1]), cnBytes[0]);
-						cn[cnBytes[0]] = '\0';
-
-						Serial.println(cn);
-
-						delete cn;
-					}
-					else
-					{
-						Serial.println(F("cert ok"));
-					}
-				}
-				else
-				{
-					Serial.println(F("cert ok"));
-				}
-			}
+			Serial.printf_P(MSG_OK);
 		}
-
-		Serial.printf_P(MSG_OK);
 	}
-	else if (!memcmp_P(&(inputBuffer[offset]), PSTR("=DELETE"), 7) && inputBufferCnt == 22)
+	// Print specific certificate
+	else if (inputBuffer[offset] == '?' && inputBufferCnt >= 16 && inputBufferCnt <= 18)
 	{
-		if (CAcert.getCount() == 0)
+		uint32_t certNumber;
+
+		++offset;
+		if (!readNumber(inputBuffer, offset, certNumber) || certNumber == 0)
 		{
-			Serial.println(F("+CIPSSLCERT:no cert"));
+			Serial.printf_P(MSG_ERROR);

```

Comment:
Why report error? I think it's fine to have no certificates.

Reply:
I had doubts about this one to. Have it report error or just ok, both is fine. The final thought I took is how would for example the WifiEspAt library handle this if this would be implemented, then I thought receiving error is easier than reading "no cert". 

But if you want it changed, I'll change it right now?

Reply:
The place where you pinned this comment is when the number read failed or the user gives 0, so in both cases more like an actual error than just no cert.

Reply:
No certificate is a valid state. The libraries using ESP and this code for a wifi connection often parse OK or ERROR to know if the command failed or succeeded. I think we should return OK with no certificate loaded as I dislike "exception driven programming".

Reply:
Sorry, the position was wrong. Added a new comment to point out the correct position.